### ISSUE TITLE
securing KubeArmor pods and policies (Policing the Police!)

### ISSUE THREAD

**Issue by nyrahul** [kubearmor/KubeArmor#1186]
KubeArmor is a security engine and thus it is imperative that it follows all the security best practices. The aim is to ensure security of the KubeArmor itself. Much of the work towards following best security practices (such as not using privileged flag) is already followed. It is now time to notch up the security practices to next level. The proposition is as follows:

- [x] Dogfooding
  - [x] Protect AppArmor policies mount point to be accessible only by KubeArmor binary. Enforce a KubeArmor policy to ensure this.
  - [x] Only allow access of service account token for kubearmor binary in the KubeArmor pod.
  - [x] Use hardening policies for KubeArmor pod.
- [x] Remove use of cluster-admin role
- [ ] Satisfy SLSA 3 requirements #1164 
- [x] Ability to use signed policy YAMLs only (orgs can leverage Adm controllers such as Kyverno)
- [x] seccomp profile for kubearmor
- [x] selinux/apparmor ootb profile
- [x] Fix all critical vulnerabilities across all kubearmor-images.
- [x] TLS for all intra-kubearmor communication #1526 
- [x] Reducing the kubearmor base image (can we try distroless kubearmor image?) [moved to UBI]
- [x] Remove unnecessary hostPaths from the manifests
- [x] Remove unnecessary capabilities from the manifests
- [x] Optimize RBAC permissions (implementation detail - config will be handled through Helm values of kubearmor-operator)
	- [x] Check if update perms can be removed for deployments/statefulsets etc - for AppArmor, patch should be suffice
	- [x] BPF LSM doesn't need the perms to annotate the above, set RBAC accordingly. (check kubearmor-policy annotation once though)
- [ ] Use `baseline` level in `enforce` mode Pod Security Admission for KubeArmor namespace
- [ ] attacker can disable kprobes, thus we won't get any future events 
        Predefined hardening policy for checking any changes to `/sys/kernel/debug/kprobes/enabled`

**Comment by Ankurk99** [comment]
[`karmor recommend` for KubeArmor Daemonset](https://gist.github.com/Ankurk99/8cc2bd7b3aebfdfd856035b3c195e3bb)

**Comment by daemon1024** [comment]
- How would KubeArmor daemonset attach a profile to itself?
  - Failed admission if annotation added and no profile found

Potential Solution
- Snitch has the intelligence to load the default apparmor profile.

**Comment by DelusionalOptimist** [comment]
For protecting KubeArmor with KubeArmor in BPF LSM, we need to:
- How to differentiate between host process and containers with hostPid.

**Comment by daemon1024** [comment]
- Seccomp Loading 
- Try running all other KA Components with Seccomp RuntimeDefault

**Comment by daemon1024** [comment]
- Remove skipping kubearmor-app logic from KubeArmor. 
- Have snitch generate apparmor profile just for kubearmor daemonset with a privileged default profile
- annotate kubearmor daemonset if snitch successful
- verify if matchlabels works as expected with kubearmor pods as well 

Edge Case
- kubearmore-controller should not block daemonset deployment or try to admission control daemonset (it can do it for other kubearmor pods except snitch)
- if operator does not exist, verify what happens when kubearmor tries to patch it's own daemonset
   - validate if kubearmor loads apparmor profile before patching

**Comment by dirsigler** [comment]
Regarding "Use baseline level in enforce mode Pod Security Admission for KubeArmor namespace", how about providing a Namespace manifest within the Helm chart which already provide the specific labels to set the Pod Security Standards on it.

This way users could for example deploy the Chart via Helm and use their own Namespace via `--namespace --create-namespace` or implement the provided namespace with all the features out of the box.

**Comment by DelusionalOptimist** [comment]
**Prioritize**

* Try using different manifests for AppArmor and BPF LSM
* Remove hostPath volumes
* Remove unnecessary capabilities from the manifests
* Have an FAQ on additional privilleges/host path volumes required if we can't get around them
* Optimize for best security for BPF LSM first
* Enable seccomp by default

**Future**
* Try using different images for AppArmor and BPF LSM
  * Use UBI image if BPF-LSM is present? (kubearmor-ubi)
      - Can't do - this doesn't work with containerd
      - Advantages - non root user not required
  * Use a more minimal image which doesn't ship with apparmor_parser or any other caps required by apparmor

**Comment by DelusionalOptimist** [comment]
### Host paths

#### KubeArmor
- Remove:
	- `/sys/fs/bpf` - not needed, KubeArmor can create and mount bpffs within the container, persistence of maps not an issue
	- `/sys/kernel/debug` - was needed earlier for bpf_printk logs, not needed anymore
	
	- Init container mounts: if BTF present, compilation not needed - remove the init container entirely along with below volumes. Also document when would this be needed. (Checkout PR #1658)
		- `/etc/os-release`
		- `/lib/modules`
		- `/usr/src`

- Document:
	- Container runtime sockets (read-only) - needed for interacting with container runtimes, would be removed soon as we adopt OCI hooks (PR in review)
	- `/etc/apparmor.d` (read/write) - needed for writing apparmor profiles. Only mounted if AppArmor enforcer is being used.
	
#### Snitch
- Remove: 
	- `/` - snitch mounts entire root, though read-only but it is not good. Mount specific paths. Also, document each of the path that snitch mounts for detection of the env, if we cannot remove them.

- Document:
	- `/var/lib/kubelet/seccomp` - (read/write) needed for storing seccomp profiles - if not desired inform about the option to disable seccomp.

#### Controller
- Remove:
	- `/sys/kernel/security` - to check if AppArmor enforcer. Handle this through snitch and remove relevant code in controller. Label/annotate the deployment and accordingly skip adding this volume if it is not AppArmor. (Checkout PR: https://github.com/kubearmor/KubeArmor/pull/1497)

**Comment by DelusionalOptimist** [comment]
### Capabilities

#### KubeArmor/InitContainer
- Remove:
	- `SYS_RESOURCE` - for managing memory. Needed only on kernel < 5.11 by cilium/ebpf [(ref)](https://github.com/cilium/ebpf/blob/65fb774468cbc51047f7115a57623832e8388eff/rlimit/rlimit.go#L103). Modify snitch to handle this.

- Check:
	- 	`SYS_ADMIN` - Our usecases should ideally be satisfied with `CAP_BPF` itself however SYS_ADMIN is still kept
		- In order to be compatible with Kernels older than 5.8.
		- Certain problems encountered in loading eBPF code with CAP_BPF (we'll try to see if we can get around these anyhow)
	- `SETUID` - trace and check
	- `SETGID` - trace and check
	- `IPC_LOCK` - for locking memory using mmap and other related syscalls as mentioned in man page. ([Ref](https://github.com/cilium/cilium/blob/ed65b5d1e5aa6b2f4be0f9ba2d9d7962cebcf6cc/install/kubernetes/cilium/values.yaml#L261-L262))

- Document:
	- `SETPCAP` - For AppArmor subprofiles
	- `SYS_PTRACE` - for reading processes other than self from procfs (source: man capabilities)
	- `MAC_ADMIN` - for managing LSM rules (man capabilities and [ref](https://patchwork.kernel.org/project/linux-security-module/patch/20191220154208.15895-14-kpsingh@chromium.org/))
	- `CAP_DAC_OVERRIDE`, `CAP_DAC_READ_SEARCH` - mainly needed by cilium/ebpf [(ref)](https://github.com/cilium/ebpf/pull/823). Also possible that these are needed for managing apparmor profiles at /etc/apparmor.d
	
	
#### Snitch
- Check and document:
	- `IPC_LOCK`
	- `SYS_ADMIN`
	- `SYS_RESOURCE`

**Comment by DelusionalOptimist** [comment]
> Regarding "Use baseline level in enforce mode Pod Security Admission for KubeArmor namespace", how about providing a Namespace manifest within the Helm chart which already provide the specific labels to set the Pod Security Standards on it.
> 
> This way users could for example deploy the Chart via Helm and use their own Namespace via `--namespace --create-namespace` or implement the provided namespace with all the features out of the box.

Yep! This seems fairly straight and user friendly. Thanks for the suggestion @dirsigler.
Also, no one is looking into this aspect right now so PRs welcome! : )

**Comment by DelusionalOptimist** [comment]
Signing container images
SBOM
PR - https://github.com/kubearmor/KubeArmor/pull/1735

**Comment by angellk** [comment]
Please complete and merge this before the KubeArmor Incubation application can be picked up for Due Diligence.

**Comment by soorya-knox** [comment]
Update the document for Securing KubeArmor

**Comment by nyrahul** [comment]
Closing this issue with the following reasoning.

* Disable kprobes ... This remains a problem wherein an attacker can disable kprobes altogether at the host layer and prevent policy violations from getting reported. Note that "Policy Enforcement" won't be impacted by this.
* salsa level 3 compliance is still a possibility and a separate issue exists for it #1164 

Multiple teams worked on this issue over multiple years to handle action items covering everything from kernel-based handling to ensuring appropriate Kubernetes security primitives are used for secure deployment of KubeArmor.

One of the biggest achievements for KubeArmor has been that it does not require a `privileged` container, unlike most other engines in the same space, even though KubeArmor provides pre-emptive mitigation options, unlike most others. This is no easy feat!
